### PR TITLE
Replace keycloak-js with oidc-spa

### PR DESCRIPTION
**Pull request by bigglesandginger**
# Description

Replace `keycloak-js` library with `oidc-spa`.

Resolves #(issue)

Primarily refreshes the token correctly
Additionally `keycloak-js` is end of life

### CODE REVIEW

**Comment by bigglesandginger** [comment]
What this won't fix is the issue that once a bearer token has expired and one has the application open in multiple windows, then at least one of the windows will be effectively logged out.  I am trying to figure out what causes this.

File: packages/client/hmi-client/src/main.ts [hunk 1]
```diff
@@ -45,16 +43,16 @@ app.use(createPinia());
...
 await authStore.init();
 logger.info('Authenticated');
 await clientEventServiceInit();
 // Token Refresh
-setInterval(async () => {
-	await window.keycloak.updateToken(70);
-}, 6000);
+// setInterval(async () => {
+// 	await window.keycloak.updateToken(70);
+// }, 6000);
```
Comment (YohannParis) [inline packages/client/hmi-client/src/main.ts:55]:
```suggestion
```

File: packages/client/hmi-client/src/main.ts [hunk 2]
```diff
@@ -45,16 +43,16 @@ app.use(createPinia());
...
 // Set up the Keycloak authentication
 const authStore = useAuthStore();
-authStore.setKeycloak(window.keycloak);
+authStore.setOidc(oidc);
 
 // Initialize user
 await authStore.init();
 logger.info('Authenticated');
 await clientEventServiceInit();
 // Token Refresh
```
Comment (YohannParis) [inline packages/client/hmi-client/src/main.ts:52]:
```suggestion
```

File: packages/client/hmi-client/src/main.ts [hunk 3]
```diff
@@ -20,22 +20,20 @@ import App from '@/App.vue';
...
-		keycloak_init: Promise<boolean>;
-		keycloak: Keycloak;
-	}
-}
+const oidcSettings = await fetch('/api/configuration/keycloak').then((r) => r.json());
+const oidc = await createOidc({
+	issuerUri: `${oidcSettings['auth-server-url']}/realms/${oidcSettings.realm}`,
+	clientId: oidcSettings.resource,
+	publicUrl: '/'
+});
```
Comment (YohannParis) [inline packages/client/hmi-client/src/main.ts:31]: Why this is not done within `useAuthStore.init` method?

**Review by YohannParis — COMMENTED** [review]
Can we remove the horrible hash in the URL once logged in?
`http://localhost:8080/?configHash=efa63722&state=c5aab631a88f4fb290678fd8bbd396a1&session_state=8e451ac0-c26b-4c21-9303-de3b269fc00f&code=34c2f703-77fc-4782-b20f-671b0cbba348.8e451ac0-c26b-4c21-9303-de3b269fc00f.316f53e1-2b9f-439b-a300-2a5d6726355f`

File: packages/client/hmi-client/src/main.ts (hunk 3 above)
Comment (bigglesandginger) [reply packages/client/hmi-client/src/main.ts:31]: A great question, before a lot was littered within `main.ts` - I will look into tiding this up

**Comment by bigglesandginger** [comment]
> Can we remove the horrible hash in the URL once logged in? `http://localhost:8080/?configHash=efa63722&state=c5aab631a88f4fb290678fd8bbd396a1&session_state=8e451ac0-c26b-4c21-9303-de3b269fc00f&code=34c2f703-77fc-4782-b20f-671b0cbba348.8e451ac0-c26b-4c21-9303-de3b269fc00f.316f53e1-2b9f-439b-a300-2a5d6726355f`


After playing with this I cannot find a way to remove the `configHash`, `state`, and `session_state` query parameters without forcing a page redirect that is broken (ie the page does not load correctly - 401s).  These are not `location.hash` parameters (which maybe can be removed without issue) but rather the location.search which when changed cause the page to reload.

File: packages/client/hmi-client/package.json [hunk 4]
```diff
@@ -56,6 +56,7 @@
 		"lodash": "4.17.21",
 		"mathlive": "0.89.2",
 		"mathml-to-latex": "1.2.0",
+		"oidc-spa": "^4.6.1",
```
Comment (dgauldie) [inline packages/client/hmi-client/package.json:59]: you should remove the ^ and pin to a specific version. our renovate bot will do the rest

File: packages/client/hmi-client/package.json [hunk 5]
```diff
@@ -86,6 +87,7 @@
 		"@types/p5": "1.6.2",
 		"@vitejs/plugin-vue": "4.6.2",
 		"@vitest/coverage-c8": "0.24.5",
+		"jsdom": "^24.1.0",
```
Comment (dgauldie) [inline packages/client/hmi-client/package.json:90]: here too

File: packages/client/hmi-client/src/main.ts [hunk 6]
```diff
@@ -20,41 +19,20 @@ import App from '@/App.vue';
 import { useProjects } from '@/composables/project';
 import { useNotificationManager } from '@/composables/notificationManager';
 import '@/assets/css/style.scss';
-import Keycloak from 'keycloak-js';
+// import { createOidc } from 'oidc-spa';
```
Comment (dgauldie) [inline packages/client/hmi-client/src/main.ts:22]: do we still need this if it's commented out?

File: packages/client/hmi-client/vite.config.ts [hunk 7]
```diff
@@ -91,6 +91,8 @@ export default defineConfig({
 		reporters: ['junit', 'default'],
 		outputFile: {
 			junit: './tests/unit/reports/junit-report.xml'
-		}
+		},
+		environment: 'jsdom'
 	}
+	// define: process.env.VITEST ? {} : { global: 'window' }
```
Comment (YohannParis) [inline packages/client/hmi-client/vite.config.ts:97]: Can be removed

**Comment by mwdchang** [comment]
Do we still need to carry `keycloak.js` as a part of the client dependency?

**Pull request merged**